tools/block-generator/generator/config.go

```diff
@@ -0,0 +1,323 @@
+	}
+
+	if cfg.NumGenesisAccounts == 0 {
+		if defaults {
+			cfg.NumGenesisAccounts = defaultGenesisAccountsCount
+		} else {
+			return fmt.Errorf("number of genesis accounts must be > 0")
+		}
+	}
+
+	if cfg.GenesisAccountInitialBalance == 0 {
+		if defaults {
+			cfg.GenesisAccountInitialBalance = defaultGenesisAccountInitialBalance
+		} else {
+			return fmt.Errorf("genesis account initial balance must be > 0")
+		}
+	}
+
+	var weights []*float32
+
+	weights = []*float32{&cfg.PaymentTransactionFraction, &cfg.AssetTransactionFraction, &cfg.AppTransactionFraction}
+	if eTxnTypes := sumIsCloseToOneWithDefault(defaults, weights...); eTxnTypes != nil {
+		return fmt.Errorf("transaction distribution ratios sum should equal 1: %w", eTxnTypes)
+	}
+
+	weights = []*float32{&cfg.PaymentNewAccountFraction, &cfg.PaymentFraction}
+	if ePymtTypes := sumIsCloseToOneWithDefault(defaults, weights...); ePymtTypes != nil {
+		return fmt.Errorf("payment configuration ratios sum should equal 1: %w", ePymtTypes)
+	}
+
+	weights = []*float32{&cfg.AssetCreateFraction, &cfg.AssetDestroyFraction, &cfg.AssetOptinFraction, &cfg.AssetCloseFraction, &cfg.AssetXferFraction}
+	if eAssetTypes := sumIsCloseToOneWithDefault(defaults, weights...); eAssetTypes != nil {
+		return fmt.Errorf("asset configuration ratios sum should equal 1: %w", eAssetTypes)
+	}
+
+	weights = []*float32{&cfg.AppSwapFraction, &cfg.AppBoxesFraction}
+	if eAppTypes := sumIsCloseToOneWithDefault(defaults, weights...); eAppTypes != nil {
+		return fmt.Errorf("app configuration ratios sum should equal 1: %w", eAppTypes)
+	}
+
+	weights = []*float32{&cfg.AppSwapCreateFraction, &cfg.AppSwapUpdateFraction, &cfg.AppSwapDeleteFraction, &cfg.AppSwapOptinFraction, &cfg.AppSwapCallFraction, &cfg.AppSwapCloseFraction, &cfg.AppSwapClearFraction}
+	if eAppSwapTypes := sumIsCloseToOneWithDefault(defaults, weights...); eAppSwapTypes != nil {
+		return fmt.Errorf("app swap configuration ratios sum should equal 1: %w", eAppSwapTypes)
+	}
+
+	weights = []*float32{&cfg.AppBoxesCreateFraction, &cfg.AppBoxesUpdateFraction, &cfg.AppBoxesDeleteFraction, &cfg.AppBoxesOptinFraction, &cfg.AppBoxesCallFraction, &cfg.AppBoxesCloseFraction, &cfg.AppBoxesClearFraction}
+	if eAppBoxesTypes := sumIsCloseToOneWithDefault(defaults, weights...); eAppBoxesTypes != nil {
+		return fmt.Errorf("app boxes configuration ratios sum should equal 1: %w", eAppBoxesTypes)
+
+	}
+
+	return nil
+}
+
+func asPtrSlice(weights []float32) []*float32 {
+	ptrs := make([]*float32, len(weights))
+	for i := range weights {
+		weight := weights[i]
+		ptrs[i] = &weight
+	}
+	return ptrs
+}
+
+// sumIsCloseToOneWithDefault returns no error if the sum of the params is close to 1.
+// It returns an error if any of the params are negative.
+// Finally, in the case that all the params are zero, it sets the first param to 1 and returns no error.
+func sumIsCloseToOneWithDefault(defaults bool, params ...*float32) error {
```

Comment:
`sumIsCloseToOneWithDefault` and `validateSumCloseToOne` replace `sumIsCloseToOne` and add the ability to default missing sections.

Reply:
Defaults is so you don't have to set anything for unused types? Maybe it could be 1 or 0?

Reply:
Why do they need to be pointers? it would be nice to avoid `asPtrSlice` everywhere.

Reply:
Because these when `defaults == true` they modify the given values so I need to pass by pointer.

Reply:
> Defaults is so you don't have to set anything for unused types? Maybe it could be 1 or 0?
When an entire section is 0-valued, and `defaults == true` the first pointer is set to 1.

Reply:
> Why do they need to be pointers? it would be nice to avoid `asPtrSlice` everywhere.

Because these when `defaults == true` they modify the given values so I need to pass by pointer.